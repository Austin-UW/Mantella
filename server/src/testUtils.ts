import { ProjectFieldsFragment } from './types'
import { MutationResult } from 'react-apollo'
import { ApolloServer } from 'apollo-server-express'
import { queries } from './graphql/queries/queries'
import { mutations } from './graphql/mutations/mutations'
import path from 'path'
import fs from 'fs'
import mongoose from 'mongoose'

// Load models since we will not be instantiating our express server.
mongoose.connect(
  process.env.DB_CONNECT ||
    'mongodb+srv://conrad:cokay101@kanbanbrawn-41gbz.mongodb.net/kanban?retryWrites=true',
  { useNewUrlParser: true }
)

mongoose.set('useFindAndModify', false)
mongoose.set('useCreateIndex', true)

const { createTestClient } = require('apollo-server-testing')

const resolvers: any = {
  Query: queries,
  Mutation: mutations
}

const typeDefs: any = [
  fs.readFileSync(path.join(__dirname, './graphql/schema.graphql'), 'utf8')
]

const gqlServer = new ApolloServer({
  typeDefs,
  resolvers,
  tracing: false,
  context: ({ req, res }) => {
    return {
      req,
      res,
      user: {
        projects: ['5d018bc58104190fe6e2ac2c'],
        _id: '5d018bc58104190fe6e2ac2d',
        password:
          '$2a$10$LL5wFKVW./hWctu8lypf/.geDiezQhQG6rYdbI7fIsVd0wfzYJwha',
        email: 'conradkay123@gmjail.com',
        username: 'kachow',
        profileImg:
          'https://mb.cision.com/Public/12278/2797280/879bd164c711a736_800x800ar.png',
        __v: 0
      }
    }
  }
})

export const test = createTestClient(gqlServer)

export const gqlReq = async <Q, V>(args: {
  query: any
  variables: V
  isMutation: boolean
  onError?: (err: any) => void
}): Promise<MutationResult<Q>> => {
  const result = args.isMutation
    ? await test.mutate({ mutation: args.query, variables: args.variables })
    : await test.query({ query: args.query, variables: args.variables })

  return result
}

export const projToSnapshot = (project: ProjectFieldsFragment) => {
  return {
    ...project,
    id: 'PROJECT_ID',
    columns: project.columns.map((column, i) => {
      return {
        ...column,
        id: `COL_ID ${i}`,
        taskIds: column.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    columnIds: project.columnOrder.map((columnId, i) => `COL_ID ${i}`),
    ownerId: 'OWNER_ID',
    swimlanes: project.swimlanes.map((swimlane, i) => {
      return {
        ...swimlane,
        id: `SWIM_ID ${i}`,
        taskIds: swimlane.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    tags: project.tags!.map((tag, i) => {
      return { ...tag, id: `TAG_ID ${i}` }
    }),
    tasks: project.tasks.map((task, i) => {
      return {
        ...task,
        id: `TASK_ID ${i}`,
        subTasks: task.subTasks.map((subTask, e) => {
          return { ...subTask, id: `SUBTASK_ID ${e}` }
        }),
        comments: task.comments.map((comment, e) => {
          return { ...comment, id: `COMMENT_ID ${e}` }
        })
      }
    }),
    users: project.users!.map((user, i) => {
      return { id: `USER_ID ${i}` }
    })
  }
}
