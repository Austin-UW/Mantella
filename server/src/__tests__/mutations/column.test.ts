import {
  GQL_CREATE_COLUMN,
  GQL_EDIT_COLUMN,
  GQL_DELETE_COLUMN
} from './../../../../client/src/graphql/mutations/column'
import {
  CreateColumnMutation,
  CreateColumnMutationVariables,
  EditColumnMutation,
  EditColumnMutationVariables,
  DeleteColumnMutation,
  DeleteColumnMutationVariables
} from './../../../../client/src/graphql/types'
import { gqlReq, projToSnapshot } from '../../testUtils'

describe('Column Mutations', () => {
  test('createColumn editColumn deleteColumn', async () => {
    const PROJECT_ID = '5d0802a03c5daa2803c62742'

    const { data: createData } = await gqlReq<
      CreateColumnMutation,
      CreateColumnMutationVariables
    >({
      isMutation: true,
      query: GQL_CREATE_COLUMN,
      variables: { projId: PROJECT_ID, name: 'column name' }
    })

    expect(createData!.createColumn).toBeTruthy()

    const col = createData!.createColumn!.column!

    expect(
      createData!.createColumn!.project!.columnOrder!.includes(col.id)
    ).toBeTruthy()
    expect(col.name).toBe('column name')
    expect(
      projToSnapshot({
        ...createData!.createColumn!.project!,
        columns: [],
        columnIds: []
      })
    ).toMatchSnapshot()

    /* edit column */

    const { data: editData } = await gqlReq<
      EditColumnMutation,
      EditColumnMutationVariables
    >({
      isMutation: true,
      query: GQL_EDIT_COLUMN,
      variables: {
        projectId: PROJECT_ID,
        id: col.id,
        newCol: {
          name: 'new name lol',
          taskLimit: 10,
          taskIds: ['lol taskId']
        }
      }
    })

    const newCol = editData!.editColumn!.column!

    expect({
      name: newCol.name,
      taskIds: newCol.taskIds,
      id: newCol.id,
      wipLimit: newCol.taskLimit
    }).toEqual({
      name: 'new name lol',
      taskIds: ['lol taskId'],
      id: col.id,
      wipLimit: 10
    })

    expect({ ...newCol, id: 'LOLCOLID' }).toMatchSnapshot()

    /* delete column */

    const { data: deleteData } = await gqlReq<
      DeleteColumnMutation,
      DeleteColumnMutationVariables
    >({
      isMutation: true,
      query: GQL_DELETE_COLUMN,
      variables: {
        projectId: PROJECT_ID,
        id: newCol.id
      }
    })

    expect(
      deleteData!.deleteColumn!.project!.columnOrder.includes(newCol.id)
    ).toBe(false)

    expect(
      deleteData!.deleteColumn!.project!.columns.findIndex(
        (coll) => coll.id === newCol.id
      )
    ).toBe(-1)

    expect(projToSnapshot(deleteData!.deleteColumn!.project!)).toMatchSnapshot()
  })
})
