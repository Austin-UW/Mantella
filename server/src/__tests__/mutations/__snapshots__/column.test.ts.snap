// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column Mutations createColumn ediTList deleteColumn 1`] = `
Object {
  "columnIds": Array [],
  "columns": Array [],
  "id": "PROJECT_ID",
  "name": "PHOON_TIME",
  "ownerId": "OWNER_ID",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [],
  "users": Array [],
}
`;

exports[`Column Mutations createColumn ediTList deleteColumn 2`] = `
Object {
  "id": "LOLCOLID",
  "isCompletedColumn": true,
  "name": "new name lol",
  "taskIds": Array [
    "lol taskId",
  ],
  "taskLimit": 10,
}
`;

exports[`Column Mutations createColumn ediTList deleteColumn 3`] = `
Object {
  "columnIds": Array [
    "COL_ID 0",
    "COL_ID 1",
  ],
  "columns": Array [
    Object {
      "id": "COL_ID 0",
      "isCompletedColumn": false,
      "name": "Todo",
      "taskIds": Array [],
      "taskLimit": 0,
    },
    Object {
      "id": "COL_ID 1",
      "isCompletedColumn": false,
      "name": "column name",
      "taskIds": Array [],
      "taskLimit": null,
    },
  ],
  "id": "PROJECT_ID",
  "name": "PHOON_TIME",
  "ownerId": "OWNER_ID",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [],
  "users": Array [],
}
`;
